File: spec/models/reconciliation_spec.rb
# frozen_string_literal: true

require 'rails_helper'

#
# Reconciliation
#
# id, int primary key
# account_id, int foreign key
# statement_date, date
# statement_balance, decimal
# reconciled, boolean
#

RSpec.describe Reconciliation do
  it 'has a valid factory' do
    r = FactoryBot.create(:reconciliation)

    expect(r).to be_valid
    expect(r).to be_a(described_class)
  end

  describe 'validations' do
    it 'is invalid without an account' do
      expect(FactoryBot.build(:reconciliation, account: nil)).not_to be_valid
    end

    it 'is invalid without a statement date' do
      expect(FactoryBot.build(:reconciliation, statement_date: nil)).not_to be_valid
    end

    it 'is invalid without a statement balance' do
      expect(FactoryBot.build(:reconciliation, statement_balance: nil)).not_to be_valid
    end
  end

  describe 'relationships' do
    it 'belongs to account' do
      a = FactoryBot.create(:account)
      expect(FactoryBot.create(:reconciliation, account: a).account).to eq(a)
    end

    it 'has many transactions' do
      r = FactoryBot.create(:reconciliation)
      FactoryBot.create(:transaction, reconciliation: r)
      FactoryBot.create(:transaction, reconciliation: r)

      expect(r.transactions.length).to eq(2)
    end
  end

  describe 'initialize' do
    it 'sets reconciled to false by default' do
      expect(FactoryBot.create(:reconciliation).reconciled).to be(false)
    end
  end
end

File: spec/javascripts/models/reconciliation_spec.js
describe('Reconciliation', function(){
  var reconciliation;
  beforeEach(function(){
    reconciliation = new MyMoney.Models.Reconciliation({id: 10, account_id: 13});
  });

  it('has a URL', function(){
    expect(reconciliation.url()).toEqual('accounts/13/reconciliations/10');
  });

  describe('validation', function(){
    it ('requires a statement date', function(){
      expect(reconciliation.preValidate('statement_date', '   ')).toEqual('Statement date is required');
      expect(reconciliation.preValidate('statement_date', '01-Jul-2014')).toEqual('');
    });

    it ('requires a statement balance', function(){
      expect(reconciliation.preValidate('statement_balance', '   ')).toEqual('Statement balance is required');
      expect(reconciliation.preValidate('statement_balance', 300)).toEqual('');
    });
  });
});